bser_parse.c

```diff
@@ -0,0 +1,311 @@
+
+#include "bser.h"
+
+#include <stdio.h>
+#include <string.h>
+#include <jansson.h>
+
+static bser_t* bser_new_unparsed(
+        bser_buffer_t* buffer, bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_UNPARSED;
+    fill->value.unparsed.buffer = buffer;
+    return fill;
+}
+
+static bser_t* bser_new_error(
+        const char* message, bser_t* fill) {
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_ERROR;
+    fill->value.error_message = message;
+    return fill;
+}
+
+static bser_t* bser_new_compact_object(
+        bser_buffer_t* buffer, bser_t* header, bser_t* fill) {
+    size_t sz = bser_array_size(header);
+    bser_key_value_pair_t* fields = (bser_key_value_pair_t*)malloc(
+        sizeof(bser_key_value_pair_t) * sz);
+
+    for (int i = 0; i < sz; ++i) {
+        fields[i].key = *(bser_array_get(header, i));
+        bser_new_unparsed(buffer, &fields[i].value);
+    }
+    bser_new_object(fields, sz, fill);
+    return fill;
+}
+
+bser_t* bser_parse(uint8_t* data, size_t buflen, bser_t* fill) {
+    if (fill == NULL) {
+        fill = (bser_t*)malloc(sizeof(bser_t));
+    }
+    bser_buffer_t* buffer = (bser_buffer_t*)malloc(sizeof(*buffer));
+    buffer->data = data;
+    buffer->datalen = buflen;
+    buffer->cursor = 0;
+
+    bser_new_unparsed(buffer, fill);
+    return fill;
+}
+
+static int integer_from_file(FILE* fp, uint64_t* value) {
+    uint8_t  v8  = 0;
+    uint16_t v16 = 0;
+    uint32_t v32 = 0;
+    uint64_t v64 = 0;
+
+    switch (fgetc(fp)) {
+        case BSER_TAG_INT8:
+            if (fread(&v8, sizeof(v8), 1, fp) == 1) {
+                *value = v8;
+                return 0;
+            }
+        case BSER_TAG_INT16:
+            if (fread(&v16, sizeof(v16), 1, fp) == 1) {
+                *value = v16;
+                return 0;
+            }
+        case BSER_TAG_INT32:
+            if (fread(&v32, sizeof(v32), 1, fp) == 1) {
+                *value = v32;
+                return 0;
+            }
+        case BSER_TAG_INT64:
+            if (fread(&v64, sizeof(v64), 1, fp) == 1) {
+                *value = v64;
+                return 0;
+            }
+    }
+    return -1;
+}
+
+bser_t* bser_parse_from_file(FILE* fp, bser_t* fill) {
```

Comment:
mark this as for testing only? It's going to be pretty slow because of fgetc instead of fgetc_unlocked.